File: php/tests/WrapTest.php
<?php

declare(strict_types=1);

namespace Xag\FlightRecorder\Tests;

use PHPUnit\Framework\TestCase;
use Xag\FlightRecorder\Recorder;
use Xag\FlightRecorder\Recording;
use Xag\FlightRecorder\Replay;

/** One row of the fake store, so replay has a declared type to fit an answer back into. */
final class Row
{
    public function __construct(public readonly string $name, public readonly int $x)
    {
    }
}

/** The real client. It counts what it was actually asked to do. */
final class RealStore
{
    public int $reads = 0;
    public int $writes = 0;

    public function read(string $key): Row
    {
        $this->reads++;
        return new Row('Alice', 3);
    }

    public function write(string $key, string $value): string
    {
        $this->writes++;
        return 'OK';
    }

    public function untouched(string $key): string
    {
        return "not recorded: $key";
    }
}

/** Wrapping a client object: the boundary is the object, as it is in Node, .NET, Go and Java. */
final class WrapTest extends TestCase
{
    use TempDir;

    public function testAWrappedCallIsForwardedToTheRealThingAndWrittenDown(): void
    {
        $real = new RealStore();
        $kv = Recorder::wrapAs('kv', $real, 'read', 'write');

        $rec = Recorder::open($this->tempDir(), Toy::plainBoundary());
        $rec->call('load', ['key' => 'alice'], static fn (): Row => $kv->read('alice'));

        self::assertSame(1, $real->reads, 'the real object must actually have been called');

        $cv = Recording::load($rec->path())->call(0);
        // Prefix-qualified, so two clients never collide on the tape.
        self::assertSame('kv.read', $cv->event('fx')['fn']);
    }

    public function testAMethodNotNamedIsInvisibleToTheRecorder(): void
    {
        $real = new RealStore();
        $kv = Recorder::wrapAs('kv', $real, 'read');

        $rec = Recorder::open($this->tempDir(), Toy::plainBoundary());
        $out = null;
        $rec->call('other', [], static function () use ($kv, &$out): string {
            $out = $kv->untouched('x');
            return 'done';
        });

        self::assertSame('not recorded: x', $out);
        self::assertNull(Recording::load($rec->path())->call(0)->event('fx'));
    }

    /** Replay must not reach the real world. */
    public function testReplayHandsBackTheDeclaredTypeNotARawArray(): void
    {
        $real = new RealStore();
        $kv = Recorder::wrapAs('kv', $real, 'read', 'write');
        $tool = static fn (): Row => $kv->read('alice');

        $rec = Recorder::open($this->tempDir(), Toy::plainBoundary());
        $rec->call('load', [], $tool);
        self::assertSame(1, $real->reads);

        // A holder, not a by-reference capture: an arrow function captures by value, so `&$x`
        // inside one binds to the arrow's copy and the assignment never reaches this scope.
        $captured = new \ArrayObject();
        $body = static function () use ($tool, $captured): Row {
            $row = $tool();
            $captured['row'] = $row;
            return $row;
        };

        $r = Replay::replay(
            $rec->path(),
            0,
            static fn (): callable => $body,
            Toy::plainBoundary()
        );

        self::assertNull($r->divergence, (string) $r);
        self::assertSame(1, $real->reads, 'replay must not reach the real world');
        self::assertInstanceOf(Row::class, $captured['row'], 'the declared type must survive the tape');
        self::assertSame('Alice', $captured['row']->name);
        self::assertSame(3, $captured['row']->x);
    }

    /** Replaying a run must not charge the card twice. */
    public function testAWriteIsRecordedAndNotReExecutedOnReplay(): void
    {
        $real = new RealStore();
        $kv = Recorder::wrapAs('kv', $real, 'read', 'write');
        $tool = static fn (): string => $kv->write('k', 'v');

        $rec = Recorder::open($this->tempDir(), Toy::plainBoundary());
        $rec->call('save', [], $tool);
        self::assertSame(1, $real->writes);

        $r = Replay::replay($rec->path(), 0, static fn (): callable => $tool, Toy::plainBoundary());
        self::assertTrue($r->ok(), (string) $r);
        self::assertSame(1, $real->writes, 'the write must not have run again');
    }

    public function testUnwrapReachesTheRealObjectForCodeThatNeedsTheDeclaredType(): void
    {
        $real = new RealStore();
        $kv = Recorder::wrapAs('kv', $real, 'read');
        self::assertSame($real, $kv->unwrap());
    }
}

File: java/src/test/java/io/github/xag/flightrecorder/WrapTest.java
package io.github.xag.flightrecorder;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Path;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.jupiter.api.Assertions.*;

/** Wrapping what the app holds: a transparent proxy, not a mock. */
class WrapTest {

    record Row(String name, int x) {}

    interface Store {
        Row read(String key);
        void write(String key, String value);
        String untouched(String key);
    }

    static final class RealStore implements Store {
        final AtomicInteger reads = new AtomicInteger();
        final AtomicInteger writes = new AtomicInteger();

        @Override public Row read(String key) {
            reads.incrementAndGet();
            return new Row("Alice", 3);
        }

        @Override public void write(String key, String value) { writes.incrementAndGet(); }

        @Override public String untouched(String key) { return "passthrough"; }
    }

    @Test
    void aWrappedCallIsForwardedToTheRealThingAndWrittenDown(@TempDir Path tmp) throws Exception {
        RealStore real = new RealStore();
        Store store = Recorder.wrapAs(Store.class, real, "kv", "read", "write");

        try (Recorder rec = Recorder.open(tmp.toString(), new Boundary())) {
            Row row = rec.call("load", Map.of(), () -> store.read("alice"));
            assertEquals(new Row("Alice", 3), row);
            assertEquals(1, real.reads.get(), "the real object was actually called");

            Recording tape = Recording.load(rec.path());
            assertEquals("kv.read", tape.call(0).event("fx").get("fn"),
                    "the prefix qualifies the name so two clients never collide on the tape");
        }
    }

    @Test
    void aMethodNotNamedIsInvisibleToTheRecorder(@TempDir Path tmp) throws Exception {
        RealStore real = new RealStore();
        Store store = Recorder.wrapAs(Store.class, real, "kv", "read");

        try (Recorder rec = Recorder.open(tmp.toString(), new Boundary())) {
            String out = rec.call("x", Map.of(), () -> store.untouched("k"));
            assertEquals("passthrough", out);

            Recording tape = Recording.load(rec.path());
            assertNull(tape.call(0).event("fx"), "an unnamed method is forwarded, not recorded");
        }
    }

    /**
     * The reason coercion exists. A tape stores structure, not types — so the recorded {@code Row}
     * comes back as a map. Without fitting it to the declared return type, the app fails on a cast
     * the recorder caused.
     */
    @Test
    void replayHandsBackTheDeclaredTypeNotARawMap(@TempDir Path tmp) throws Exception {
        RealStore real = new RealStore();
        Store store = Recorder.wrapAs(Store.class, real, "kv", "read", "write");

        try (Recorder rec = Recorder.open(tmp.toString(), new Boundary())) {
            rec.call("load", Map.of(), () -> store.read("alice"));

            Recording tape = Recording.load(rec.path());
            Replay.Resolver code = (fn, kwargs) -> () -> store.read("alice");
            Replay.Report r = Replay.replayCall(tape.call(0), code, new Boundary(), false);

            assertTrue(r.ok(), () -> Replay.format(0, r));
            // The real store was called once (during record) and NOT during replay.
            assertEquals(1, real.reads.get(), "replay must not reach the real world");
        }
    }

    @Test
    void aWriteIsRecordedAndNotReExecutedOnReplay(@TempDir Path tmp) throws Exception {
        RealStore real = new RealStore();
        Store store = Recorder.wrapAs(Store.class, real, "kv", "read", "write");

        try (Recorder rec = Recorder.open(tmp.toString(), new Boundary())) {
            rec.call("save", Map.of(), () -> store.write("alice", "v"));
            assertEquals(1, real.writes.get());

            Recording tape = Recording.load(rec.path());
            Replay.Resolver code = (fn, kwargs) -> () -> { store.write("alice", "v"); return null; };
            Replay.replayCall(tape.call(0), code, new Boundary(), false);

            // Replaying a run must not charge the card twice.
            assertEquals(1, real.writes.get(), "the write must not happen again on replay");
        }
    }

    @Test
    void coercionFitsRecordsNumbersAndEnums() {
        assertEquals(new Row("Alice", 3),
                Serial.coerce(Map.of("name", "Alice", "x", 3L), Row.class));
        assertEquals(3, Serial.coerce(3L, int.class));
        assertEquals(3.0, Serial.coerce(3L, double.class));
        // Best-effort: an unfittable shape comes back as it is rather than throwing.
        assertEquals("nope", Serial.coerce("nope", Row.class));
    }
}
